core/lib/pubs.ts

```diff
@@ -89,14 +92,124 @@ export const processedPubsToPubsResult = (pubs: InputPub[]): GetPubsResult => {
+export const valuesWithoutTitle = <
+	T extends ProcessedPubWithForm<{
+		withRelatedPubs: true;
+		withStage: true;
+		withPubType: true;
+		withMembers: true;
+	}>,
+>(
+	pub: T
+): T["values"] => {
 	const titleField = getTitleField(pub);
 	if (!titleField) {
 		return pub.values;
 	}
 	return pub.values.filter((value) => value.fieldId !== titleField?.id);
 };
+
+/**
+ * Merges a pub with a form so that each pub value also has information from its
+ * corresponding form element. This is a full join, such that we will also have
+ * form elements that do not have pub values, and pub values that do not have form elements.
+ */
+export const getPubByForm = ({
+	pub,
+	form,
+	withExtraPubValues,
+}: {
+	pub: FullProcessedPub;
+	form: Form;
+	withExtraPubValues: boolean;
+}): ProcessedPubWithForm<{
+	withRelatedPubs: true;
+	withStage: true;
+	withPubType: true;
+	withMembers: true;
+}> => {
+	const { values } = pub;
+	if (!values.length) {
+		return pub;
+	}
+
+	const valuesByFieldSlug = values.reduce<Record<string, FullProcessedPub["values"][number][]>>(
+		(acc, value) => {
+			if (!acc[value.fieldSlug]) {
+				acc[value.fieldSlug] = [value];
+			} else {
+				acc[value.fieldSlug].push(value);
+			}
+			return acc;
+		},
+		{}
+	);
+
+	const pubFieldFormElements = form.elements.filter((fe) => fe.type === ElementType.pubfield);
+	const valuesWithFormElements: ProcessedPubWithForm<{
+		withRelatedPubs: true;
+		withStage: true;
+		withPubType: true;
+		withMembers: true;
+	}>["values"] = [];
+	for (const formElement of pubFieldFormElements) {
+		const values = valuesByFieldSlug[formElement.slug];
+		const formInfo = {
+			formElementId: formElement.id,
+			formElementLabel: formElement.label,
+			formElementConfig: formElement.config,
+		};
+		if (!values) {
+			valuesWithFormElements.push({
+				id: null,
+				value: null,
+				createdAt: null,
+				updatedAt: null,
+				schemaName: formElement.schemaName,
+				fieldId: formElement.fieldId as PubFieldsId,
+				fieldSlug: formElement.slug,
+				fieldName: formElement.fieldName,
+				relatedPubId: null,
+				...formInfo,
+			});
+		} else {
+			for (const value of values) {
+				valuesWithFormElements.push({ ...value, ...formInfo });
+			}
+		}
+	}
+
+	if (!withExtraPubValues) {
+		return { ...pub, values: valuesWithFormElements };
+	}
+
+	const formElementSlugs = new Set(pubFieldFormElements.map((fe) => fe.slug));
+	const valueFieldSlugs = new Set(Object.keys(valuesByFieldSlug));
+	const slugsNotInForm = Array.from(valueFieldSlugs.difference(formElementSlugs));
+	const valuesNotInForm: ProcessedPubWithForm<{
+		withRelatedPubs: true;
+		withStage: true;
+		withPubType: true;
+		withMembers: true;
+	}>["values"] = [];
+	for (const slug of slugsNotInForm) {
+		for (const value of valuesByFieldSlug[slug]) {
+			valuesNotInForm.push(value);
+		}
+	}
+
+	const newValues = [...valuesWithFormElements, ...valuesNotInForm];
+
+	return { ...pub, values: newValues };
```

Comment:
nice work! reading this function makes me appreciate how terse sql is, but i think we'll be glad you did it this way!